0027-tcp-load-balancer.md

```diff
@@ -0,0 +1,264 @@
+Server Groups are an arbitary number of servers that are grouped for security aurhorisation reasons.
+Clients can only connect to one server group, this will be hardcoded as part of the authorisation code.
+The Clients key will be their public key stored as a string.
+
+#### Client Structure
+```rust
+struct Client
+{
+    public_key  : String,
+    connections : Vec<Connection>,
+}
+```
+
+This struct will house all data required to handle all connections (downstream and upstream) on a per client basis.
+It will store the client public key for the client as a means of identification.
+
+```rust
+enum ConnState
+{
+    INIT,
+    UP_CONNECTED,
+    OKAY,
+    UP_DISCONNECT,
+    UP_TIMEOUT,
+    DOWN_DISCONNECT,
+    DOWN_TIMEOUT,
+    DOWN_ENC_ERR,
+}
+```
+
+The state of both connections will be managed by a finite state machine.
+Any state after OKAY will result in the whole connection being dropped, the client will need to retry that connection.
+
+These states could also be used to aid the HealthChecker in marking a node as unhealthy (in the case of UP_DISCONNECTED or UP_TIMEOUT).
+
+The DOWN_ENC_ERR is a catchall for any and all rustls::Stream errors.
+
+```rust
+struct Connection
+{
+    down_stream : TcpStream,
+    up_stream   : TcpStream,
+    tls         : rustls::Stream, 
+    conn_state  : ConnState,
+    up_sg       : u32
+    up_id       : u32,
+}
+```
+
+The connection structure will house all TcpStreams and the rustls::Stream the id of the upstream server and the connection state enum.
+
+#### Client Rate Limiter
+
+The main library will contain a hashmap of clients with the connections that they store.
+each new connection will be first checked against the current count if it is less than a pre determined limit then
+the connection is placed throught the health and least connection forwarder and data is sent on accrdingly.
+
+```rust
+impl Client
+{
+    fn add_connection(&mut self, down_stream: TcpStream, rustls::Stream, up_id: u32, up_address: String) -> Result<(), Box<dyn std::error::Error>>
+    {
+        if self.connections.len() >= CLIENT_LIMIT
+        {
+            return Err("Client rate limit hit".into());
+        }
+
+        ... // connect to upstream
+
+        self.connections.push(Connection {...});
+
+        Ok(())
+    }
+}
+```
+
+#### Least Connections Forwarder
+
+Once the client has been authenticated and authorised for a server group, the least connection forwarder can forward the client to 
+an appropriate server.
+The ServerGroup struct will house connection stats and addresser for each server.
+
+Connections will be incremented and decremented as connections are created and destroyed by clients.
+
+As such there will be an simple API that will increment or decrement the server's connection count.
+
+```rust
+struct ServerGroup
+{
+    id           : u32,
+    server_addrs : HashMap<u32, String>,
+    cxn_cntr     : HasnMap<u32, usize>,
+}
+
+impl ServerGroup
+{
+    fn find_min(&self) -> Option<u32>
```

Comment:
Won't this only ever end up using the first id returned by `self.server_addrs.iter()`, as nothing will ever access the rest of `self.server_addrs`?

Reply:
I take it you mean if there for lines 145 through 152?

If that is the case: No. if there are no connections in the load balancer you return the first server you find.

otherwise you look for the smallest connection count in lines: 154 through 163.